pkg/repository/manifest/repo.go

```diff
@@ -57,7 +78,75 @@ func Init(dst string, initTime time.Time) error {
 		return fmt.Errorf("manifest '%s' not initialized porperly", ty)
 	}
 
-	return BatchSaveManifests(dst, manifests)
+	return BatchSaveManifests(dst, manifests, pubKey, privKey)
+}
+
+// AddComponent adds a new component to an existing repository
+func AddComponent(id, name, desc, owner, repo string, isDefault bool, pub, priv string) error {
+	// read key files
+	privBytes, err := ioutil.ReadFile(priv)
+	if err != nil {
+		return err
+	}
+	privKey := &crypto.RSAPrivKey{}
+	if err = privKey.Deserialize(privBytes); err != nil {
+		return err
+	}
+	pubBytes, err := ioutil.ReadFile(pub)
+	if err != nil {
+		return err
+	}
+	pubKey := &crypto.RSAPubKey{}
+	if err = pubKey.Deserialize(pubBytes); err != nil {
+		return err
+	}
+
+	// read manifest files from disk
+	manifests, err := ReadManifestDir(repo)
+	if err != nil {
+		return nil
+	}
+
+	// check id conflicts
+	if _, found := manifests[ManifestTypeIndex].(*Index).Components[id]; found {
+		return fmt.Errorf("component id '%s' already exist, please set another one", id)
+	}
+
+	// create new component manifest
+	currTime := time.Now().UTC()
+	comp := NewComponent(name, desc, currTime)
+
+	// update repository
+	bytes, err := json.Marshal(comp)
+	if err != nil {
+		return err
+	}
+	compInfo := ComponentItem{
+		Name:      comp.Name,
+		Owner:     owner,
+		URL:       fmt.Sprintf("/%s", comp.Filename()),
+		Length:    uint(len(bytes)),
+		Threshold: 1, // TODO: make this configurable
+	}
+	index := manifests[ManifestTypeIndex].(*Index)
+	index.Components[id] = compInfo
+	if isDefault {
+		index.DefaultComponents = append(index.DefaultComponents, id)
+	}
+	index.Version += 1 // bump index version
+
+	// update snapshot
+	snapshot := manifests[ManifestTypeSnapshot].(*Snapshot).SetVersions(manifests)
```

Comment:
I think it is better to explicitly update the version for the index and component entries, not all manifests